libraries/freertos_plus/standard/tls/src/iot_tls.c

```diff
@@ -881,6 +881,18 @@ BaseType_t TLS_Connect( void * pvContext )
 
         /* Configure the SSL context for the device credentials. */
         xResult = prvInitializeClientCredential( pxCtx );
+
+        if( xResult != 0 )
+        {
+            /* Could not initialize client credentials.
+             * This will cause client verification to fail when
+             * connecting to servers that require mutual authentication.
+             * The error is removed here, to continue with the TLS handshake.
+             * The handshake will fail if the server expects the client to
+             * verify it's identity. If only the server needs to verified,
+             * the handshake will pass. */
+            xResult = 0;
+        }
```

Comment:
If there are actual errors from `mbedtls_ssl_conf_own_cert` and other functions, won't those errors also be ignored?

Reply:
Yes but the behavior does not change. The failure will be differed to the handshake. If we fail before `mbedtls_ssl_conf_own_cert` we never add the key and certificate contexts, if we fail in `mbedtls_ssl_conf_own_cert` the behavior is identical, as they are never added to the linked list of credentials.

Reply:
How about providing a config instead that defaults to requiring client credentials? In that way, failures occurring in `prvInitializeCredentials` can be caught early by developers instead of having to rely on TLS handshake.

Reply:
I think a config is a reasonable approach. In my eyes the pros of the current changeset is it does not add any new flags or config, and introduces support with a minimal change. Already there is an explosion of configurations and I think here we can provide a reasonable default, the con being losing CPU cycles to a TLS handshake that should not have happened.

One issue I have with a build flag is that the choice is static. My guess would be that most users would like to use a basic TLS connection as well as a mutual auth connection, and would like to toggle during runtime which they woud like. 

A third approach would be to add a runtime flag for basic TLS. This would require some major API changes as well, as this layer is generally nested under other abstractions (eg. network manager, secure sockets, etc).

Reply:
>Already there is an explosion of configurations and I think here we can provide a reasonable default, the con being losing CPU cycles to a TLS handshake that should not have happened.

Yeah, that's the tradeoff with adding a config. However, I think that ignoring failures either in obtaining device credentials from PKCS11 and configuring them in mbedTLS is not desirable for debugging experience.

>One issue I have with a build flag is that the choice is static. My guess would be that most users would like to use a basic TLS connection as well as a mutual auth connection, and would like to toggle during runtime which they woud like.

With the config approach, if the user wants to support both basic TLS and mutual auth, the user can unset the config value. Wouldn't that work?

Reply:
Another issue I have with creating a config is I am not sure where it belongs. It seems wasteful to create a `iot_tls_config.h` file to contain just one configuration.

Should we place it in one of these two config files?
```c
#include "FreeRTOSIPConfig.h"
#include "core_pkcs11_config.h"
```

Reply:
Yeah, it would be wasteful imo too to create a `iot_tls_config.h` file. I will be fine place it in `FreeRTOSIPConfig.h` file

Reply:
I am not really sure if it would be a good idea to have it statically configured during compile time either. I think a runtime flag is the best choice we have so far but would likely require a lot more changes. Although I would like to emphasize the importance of not having to change any of the public-facing function signatures, maybe just having an extra `flags` parameter under `TLSContext_t`?

Reply:
`TLSContext_t` is an internal struct, but `TLSParams_t` could be modified to carry such a flag. This would also require modifying the secure_sockets interface as well, since it abstracts `iot_tls.h`.